### PR TITLE
Ensure we fire events on the hosts scheduler.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
That way we can preserve the invariant that workpsace events all fire on
the UI thread in VS.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide This is ready for review now.  This is the preferred way to do async in the workspace, while still making sure that events fire on VS' UI thread.

The reason the previous submission failed was that the MiscellaneousFilesWorkspace wasn't being created on teh UI thread, and thus all the events it fired were on the threadpool.  This was highly undesirable and meant it behaved differently from the main VS workspace.

**Comment by CyrusNajmabadi** [comment]
retest prtest/win/vsi/p1 please

**Comment by Pilchie** [comment]
I don't understand how the `MiscellaneousFilesWorkspace` wasn't already being created on the UI thread (here)[http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/LanguageService/AbstractPackage%25602.cs,59].  What is the callstack/set of user actions that triggers that?

File: src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs [hunk 1]
```diff
@@ -395,26 +395,39 @@ private ISet<DocumentId> GetProjectOpenDocuments_NoLock(ProjectId project)
...
             DocumentId documentId, SourceTextContainer textContainer,
             bool isCurrentContext = true)
         {
-            OnDocumentOpened(documentId, textContainer, isCurrentContext);
-            return SpecializedTasks.EmptyTask;
+            OnDocumentOpenedAsync(documentId, textContainer, isCurrentContext, wait: true).Wait(CancellationToken.None);
         }
 
-        protected internal void OnDocumentOpened(
+        internal Task OnDocumentOpenedAsync(
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs]: `OnDocumentOpenedAsync` only seems to be reference by tests at the moment, which is a little scary that all of our tests go through the async path, but all of the productions scenarios go through the sync path.  Is the plan to change the product code in the near future?  If not, I would prefer to remove the async path.

Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs]: Yes.  The goal is to move everyone to the async codepath.  I can make that change now as well if you want.

**Comment by CyrusNajmabadi** [comment]
> I don't understand

This was happening in RoslynPackage (which doesn't derive from AbstractPackage).  When that loads it has a ```Task.Run(() => LoadComponentsBackground());```  LoadComponentsInBackground would then call ```this.ComponentModel.GetService<MiscellaneousTodoListTable>();```.  The MiscTodoList was mef importing the misc workspace.  So the MiscWorkspace would load in the BG.

This was *never* what we wanted (for the reasons i gave above).  So i'm moving it to package initialization so that it will be initialized on hte right thread.

**Comment by Pilchie** [comment]
Got it.  I don't think there's ever a situation where we could load `RoslynPackage` without also loading an `AbstractPackage`, but there was a race there with that BG task.  Tagging @jasonmalinowski to make sure my understanding is correct.

File: src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs (hunk 1 above)
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs]: It doesn't have to be part of this PR, I just don't want us to get complacent with unit tests testing a completely different path through here than product code.

**Comment by Pilchie** [comment]
:+1:

**Comment by CyrusNajmabadi** [comment]
> Got it. I don't think there's ever a situation where we could load RoslynPackage without also loading an AbstractPackage, but there was a race there with that BG task. 

I believe that is correct.

File: src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs]: Fair enough.  :)

**Pull request merged**